File: javascript/sort/quickSort.js
const swap = (arr, i1, i2) => ([arr[i1], arr[i2]] = [arr[i2], arr[i1]]);

/**
 * 
 * 
 * @param {Array} arr 
 * @param {number} start 
 * @param {number} end 
 * @return index of the pivot
 */
function pivot(arr, start = 0, end = arr.length - 1) {
  let pivot = arr[start];
  let swapIndex = start;

for (let i = start + 1; i <= end; i++){
    if (pivot > arr[i]) {
      swapIndex++;
      swap(arr, swapIndex, i);
    }
  }

  swap(arr, start, swapIndex);
  return swapIndex;
}

// console.log(pivot([4, 8, 2, 1, 5, 7, 6, 3]));

function quickSort(arr, left = 0, right = arr.length - 1) {
  if (left < right) {
    let pivotIndex = pivot(arr, left, right);
    // left 
    quickSort(arr, left, pivotIndex - 1);
    // right
    quickSort(arr, pivotIndex+1, right);
  }
  return arr

}
console.log(quickSort([4, 8, 2, 1, 5, 7, 6, 3]));

File: python/sort/quickSort.py

def quick_sort(arr: list) -> list:
    print(arr)
    if len(arr) == 1:
        return arr
    else:
        pivot = arr[0]

        left = [item for item in arr[1:] if item < pivot]
        right = [item for item in arr[1:] if item >= pivot]

        return quick_sort(left) + [pivot] + quick_sort(right)


a = [10, 5, 2, 3, 10]
print(quick_sort(a))
